### DISCUSSION TITLE
How to Manually Set Up HotChocolate with a .NET Core API Controller?

### DISCUSSION THREAD

**Discussion by tienlx97** [ChilliCream/hotchocolate#7367] [Q&A]
Hi everyone,

I'm currently working on integrating GraphQL into an existing ASP.NET Core project, and I'm exploring how to manually set up HotChocolate with a .NET Core API controller.

### Background:

My project is structured primarily around traditional RESTful controllers, but I'd like to add GraphQL capabilities without moving away from this approach. Instead of using the default HotChocolate endpoint setup, I want to manually control the GraphQL request handling within my existing controllers.

### What I'm Trying to Achieve:

- Manual Setup: I'd like to manually configure HotChocolate in Program.cs or Startup.cs.
- Custom Controller: I want to handle GraphQL requests in a custom controller where I can directly manage the request lifecycle, including parsing the query, executing it, and returning the result.

### This is my error code:

```cs
using HotChocolate.Language;

using Microsoft.AspNetCore.Mvc;

namespace Inventory.Contracts.Graphql;

public class MyGraphQLRequest
{
  [FromHeader(Name = "X-Employee-Id")]
  public Guid EmployeeId { get; set; }

  [FromBody]
  public GraphQLRequest GraphQLRequest { get; set; } = null!;
}

public record GraphQLRequest(
  string Query,
  string? QueryId,
  string? QueryHash,
  string? OperationName,
  IReadOnlyList<IReadOnlyDictionary<string, object?>>? Variables,
  IReadOnlyDictionary<string, object?>? Extensions);

```

```cs
using HotChocolate.Execution;
using HotChocolate.Language;

using Inventory.Contracts.Graphql;
using Inventory.Infrastructure.Utils;

using Microsoft.AspNetCore.Mvc;

using HotGraphQLRequest = HotChocolate.Language.GraphQLRequest;

namespace Inventory.Api.Controllers;

[Route("/api/graphql")]
public class GraphQLController(IRequestExecutorResolver executor) : ApiController
{
  private readonly _executor = executor.GetRequestExecutorAsync().Result;

  [HttpPost]
  public async Task<IActionResult> ExecuteGraphQL([FromRequest] MyGraphQLRequest request)
  {
    if (string.IsNullOrEmpty(request.GraphQLRequest.Query))
    {
      return BadRequest("Query cannot be null or empty.");
    }

    // var operationRequest = new OperationRequest()
    // {

    // };

    // var result = _executor.ExecuteAsync(s => {

    // });

    // DocumentNode queryDocument = Utf8GraphQLParser.Parse(request.GraphQLRequest.Query);

    // var op = new OperationRequest(
    //   queryDocument,
    //   request.GraphQLRequest.OperationName,
    //   request.GraphQLRequest.Variables);

    // var queryRequest = new HotGraphQLRequest(
    //   query: queryDocument,
    //   queryHash: request.GraphQLRequest.QueryHash,
    //   queryId: request.GraphQLRequest.QueryId,
    //   operationName: request.GraphQLRequest.OperationName,
    //   variables: request.GraphQLRequest.Variables,
    //   extensions: request.GraphQLRequest.Extensions);

    // IExecutionResult result = await _executor.ExecuteAsync(queryRequest);

    // return Ok(result.ToDictionary());

    return Ok(request);
  }
}
```

**Comment by byme8** [comment]
You can just enable graphql in the existing asp net core project. Checkout my article https://dev.to/byme8/convert-rest-api-to-graphql-api-4g38 and associated repository https://github.com/byme8/DemoGraphQL

In repository has two branches:
- [master](https://github.com/byme8/DemoGraphQL/tree/master): initial rest api
- [graphql](https://github.com/byme8/DemoGraphQL/tree/graphql): modified project with graphql

[Diff](https://github.com/byme8/DemoGraphQL/compare/master...graphql)
